src/github.com/matrix-org/dendrite/clientapi/readers/directory.go

```diff
@@ -62,10 +86,86 @@ func DirectoryRoom(
 			// TODO: Return 504 if the remote server timed out.
 			return httputil.LogThenError(req, err)
 		}
+	}
 
+	return util.JSONResponse{
+		Code: 200,
+		JSON: resp,
+	}
+}
+
+// SetLocalAlias implements PUT /directory/room/{roomAlias}
+// TODO: Check if the user has the power level to set an alias
+func SetLocalAlias(
+	req *http.Request,
+	device *authtypes.Device,
+	alias string,
+	cfg *config.Dendrite,
+	aliasAPI api.RoomserverAliasAPI,
+) util.JSONResponse {
+	_, domain, err := gomatrixserverlib.SplitID('#', alias)
+	if err != nil {
 		return util.JSONResponse{
-			Code: 200,
-			JSON: resp,
+			Code: 400,
+			JSON: jsonerror.BadJSON("Room alias must be in the form '#localpart:domain'"),
 		}
 	}
+
+	if domain != cfg.Matrix.ServerName {
+		return util.JSONResponse{
+			Code: 403,
+			JSON: jsonerror.Forbidden("Alias must be on local homeserver"),
+		}
+	}
+
+	var r roomID
```

Comment:
I'd inline the struct definition here so just write:
```go
var r struct {
    RoomID `json:"room_id"`
}
```

That way I'd avoid having to think of a good name for the struct type :)
(calling it "roomID" is confusing because it's a struct containing a room ID rather than being a room ID)
If we really want it to be a named type then I'd call it something like "setAliasRequest".